modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/DatasetFragmentRenderer.java

```diff
@@ -0,0 +1,334 @@
+
+			ObjectDefinition fdsViewObjectDefinition =
+				_objectDefinitionLocalService.fetchObjectDefinition(
+					fragmentEntryLink.getCompanyId(), "C_FDSView");
+
+			if (externalReferenceCode != StringPool.BLANK) {
+				fdsViewObjectEntry = _getObjectEntry(
+					fragmentEntryLink.getCompanyId(), externalReferenceCode,
+					fdsViewObjectDefinition);
+			}
+
+			if ((fdsViewObjectEntry == null) &&
+				fragmentRendererContext.isEditMode()) {
+
+				printWriter.write(
+					StringBundler.concat(
+						"<div class=\"portlet-msg-info\">",
+						_language.get(
+							httpServletRequest, "select-a-dataset-view"),
+						"</div>"));
+			}
+
+			if (fdsViewObjectEntry == null) {
+				return;
+			}
+
+			printWriter.write(
+				_buildFragmentHTML(
+					fdsViewObjectEntry, fdsViewObjectDefinition,
+					fragmentRendererContext, httpServletRequest));
+		}
+		catch (Exception exception) {
+			throw new IOException(exception);
+		}
+	}
+
+	private String _buildFragmentHTML(
+			ObjectEntry fdsViewObjectEntry,
+			ObjectDefinition fdsViewObjectDefinition,
+			FragmentRendererContext fragmentRendererContext,
+			HttpServletRequest httpServletRequest)
+		throws Exception {
+
+		StringBundler sb = new StringBundler(5);
+
+		sb.append("<div id=\"");
+		sb.append(fragmentRendererContext.getFragmentElementId());
+		sb.append("\" >");
+
+		ComponentDescriptor componentDescriptor = new ComponentDescriptor(
+			"{FrontendDataSet} from frontend-data-set-web",
+			fragmentRendererContext.getFragmentElementId(), null, true);
+
+		Writer writer = new CharArrayWriter();
+
+		FragmentEntryLink fragmentEntryLink =
+			fragmentRendererContext.getFragmentEntryLink();
+
+		Map<String, Object> fdsViewObjectEntryProperties =
+			fdsViewObjectEntry.getProperties();
+
+		String fdsEntryObjectEntryERC = String.valueOf(
+			fdsViewObjectEntryProperties.get(
+				"r_fdsEntryFDSViewRelationship_c_fdsEntryERC"));
+
+		ObjectDefinition fdsEntryObjectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				fragmentEntryLink.getCompanyId(), "C_FDSEntry");
+
+		ObjectEntry fdsEntryObjectEntry = _getObjectEntry(
+			fragmentEntryLink.getCompanyId(), fdsEntryObjectEntryERC,
+			fdsEntryObjectDefinition);
+
+		_reactRenderer.renderReact(
+			componentDescriptor,
+			HashMapBuilder.<String, Object>put(
+				"apiURL", _getAPIURL(fdsEntryObjectEntry)
+			).put(
+				"id", "FDS_" + fragmentRendererContext.getFragmentElementId()
+			).put(
+				"namespace", fragmentRendererContext.getFragmentElementId()
+			).put(
+				"style", "fluid"
+			).put(
+				"views",
+				_getViewsJSONArray(fdsViewObjectDefinition, fdsViewObjectEntry)
+			).build(),
+			httpServletRequest, writer);
+
+		sb.append(writer.toString());
+
+		sb.append("</div>");
+
+		return sb.toString();
+	}
+
+	private String _getAPIURL(ObjectEntry fdsEntryObjectEntry) {
+		Map<String, Object> properties = fdsEntryObjectEntry.getProperties();
+
+		StringBundler sb = new StringBundler(3);
+
+		sb.append("/o");
```

Comment:
We should probably use [AbsolutePortalURLBuilderFactory](https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-url-builder/portal-url-builder-api/src/main/java/com/liferay/portal/url/builder/AbsolutePortalURLBuilderFactory.java#L33) to compose these URLs. Otherwise, they may fail when using CDNs, a root web context, etc.

Currently there's no method to compose URLs to remote APIs in [AbsolutePortalURLBuilder](https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-url-builder/portal-url-builder-api/src/main/java/com/liferay/portal/url/builder/AbsolutePortalURLBuilder.java), but we may use this PR to create it.

What do you think @dsanz @markocikos ?

Reply:
We can do it in a follow up if we don't want to delay this.

Reply:
we  should give some thought to this. IINM all "traditional" FDS usages provide a relative API url, so basically, we are delegating URL composition (everything after the port and before the given API url) [to the FDS](
https://github.com/liferay/liferay-portal/blob/f61edb7777ee566127cd0879e4a54fc7603[…]ta-set-web/src/main/resources/META-INF/resources/utils/index.js)

Reply:
I'm looking for small tasks for the new hires. This one is super nice for that purpose :)